File: src/java/main/java/composing/methods/AmountCalculator.java
package composing.methods;

public class AmountCalculator {
    public static double calculatePrice(double amount, int age){
        double discountBasedOnAge = 0;

        if(age <= 16){
            discountBasedOnAge = 0.35 * amount;
        } else if(age >= 60){
            discountBasedOnAge = 0.2 * amount;
        }
        return amount - discountBasedOnAge;
    }

    public static double calculatePrice(Order order, boolean applyAgeDiscount, int age){
        double result = 0.0;
        double discount = 0.0;
        double resultWithDiscount = 0.0;

        for (Product product: order.getProducts()) {
            result += product.getPrice();
        }

        if(applyAgeDiscount) {
            double discountBasedOnAge = 0;
            if (age <= 16) {
                discountBasedOnAge = 0.35 * result;
            } else if (age >= 60) {
                discountBasedOnAge = 0.2 * result;
            }
            discount = discountBasedOnAge;
        }
        resultWithDiscount = result - discount;

        return resultWithDiscount;
    }
}

File: src/scala/src/main/scala/composing/methods/AmountCalculator.scala
package composing.methods

object AmountCalculator {
  private val YoungMaximumAge = 16
  private val YoungDiscount = 0.35
  private val OldMinimumAge = 60
  private val OldDiscount = 0.2

  def calculatePrice(amount: Double, age: Int): Double = {
    calculate(amount, age)
  }

  def calculatePrice(
      order: Order,
      applyAgeDiscount: Boolean,
      age: Int
  ): Double = {
    calculate(order.totalPrice, age, applyAgeDiscount)
  }

  private def calculate(
      amount: Double,
      age: Int,
      applyAgeDiscount: Boolean = true
  ): Double = {
    val discount = if (applyAgeDiscount) calculateDiscountBasedOnAge(amount, age) else 0
    amount - discount
  }

  private def calculateDiscountBasedOnAge(amount: Double, age: Int): Double = {
    age match {
      case young if young <= YoungMaximumAge => YoungDiscount * amount
      case old if old >= OldMinimumAge       => OldDiscount * amount
      case _                                 => 0
    }
  }
}
